osquery/tables/system/darwin/battery.mm

```diff
@@ -0,0 +1,192 @@
+/*
+ *
+ *  Copyright (c) 2018-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under both the Apache 2.0 license (found in the
+ *  LICENSE file in the root directory of this source tree) and the GPLv2 (found
+ *  in the COPYING file in the root directory of this source tree).
+ *  You may select, at your option, one of the above-listed licenses.
+ */
+
+#include <CoreFoundation/CoreFoundation.h>
+#include <Foundation/Foundation.h>
+
+#include <osquery/tables.h>
+
+#import <IOKit/ps/IOPSKeys.h>
+#import <IOKit/ps/IOPowerSources.h>
+#import <IOKit/pwr_mgt/IOPM.h>
+#import <IOKit/pwr_mgt/IOPMLib.h>
+
+namespace osquery {
+namespace tables {
+
+NSDictionary* getIopmBatteryInfo() {
+  CFTypeRef info = IOPSCopyPowerSourcesInfo();
+
+  if (info == NULL)
+    return NULL;
+
+  CFArrayRef list = IOPSCopyPowerSourcesList(info);
+
+  // Release if there are not viable battery sources
+  if (list == NULL || !CFArrayGetCount(list)) {
+    if (list) {
+      CFRelease(list);
+    }
+    CFRelease(info);
+    return NULL;
+  }
+
+  // Assumes the first battery returned in the array is the internal battery
+  CFDictionaryRef battery = CFDictionaryCreateCopy(
+      NULL,
+      IOPSGetPowerSourceDescription(info, CFArrayGetValueAtIndex(list, 0)));
+
+  CFRelease(list);
+  CFRelease(info);
+
+  return (__bridge_transfer NSDictionary*)battery;
+}
+
+NSDictionary* getIopmpsBatteryInfo() {
+  CFMutableDictionaryRef matching, properties = NULL;
+  io_registry_entry_t entry = 0;
+  matching = IOServiceNameMatching("AppleSmartBattery");
+  entry = IOServiceGetMatchingService(kIOMasterPortDefault, matching);
+  IORegistryEntryCreateCFProperties(entry, &properties, NULL, 0);
```

Comment:
I'm not sure why you're leaks check showed no leaks, this object needs to be freed:

https://developer.apple.com/documentation/iokit/1514310-ioregistryentrycreatecfpropertie

Edit: Nevermind as ownership of this object is transferred to ARC with the bridge_transfer below. Leaving the comment here so other people will not think the same thing I did.